### PR TITLE
Basic Mob Carp Part III: Nuclear Operatives

### PR DESCRIPTION
**Pull request by Jacquerel**
## About The Pull Request

Third atomisation of #71421 
I'll stop there until these three are dealt with, before opening more.

This PR extracts all of the behaviour on Cayenne which let her be a pseudo-nuclear-operative into a component which can be used by any mob to give it the capabilities of a nuclear operative.

https://user-images.githubusercontent.com/7483112/203334848-57a6c38d-2e4f-406d-9b49-8106d31b7930.mp4
I recorded this in a way which did not pick up the UI popups so imagine the strip and numeral entry windows yourself.

Nuclear bomb operators can:
- Strip people.
- Pick up the disk.
- Put down the disk.
- Insert the disk into the bomb.
- Activate the bomb.

## Why It's Good For The Game

Moves code from being specific to one mob to generic to any mob. 
I am not sure we'll make a lot of mobs which attach this on initialise, but I am sure admins will have fun with it.

## Changelog
:cl:
refactor: Cayenne's ability to participate in Nuclear Operations is now a component which any mob can use.
/:cl:

### CODE REVIEW

File: code/datums/components/nuclear_bomb_operator.dm [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	RegisterSignal(parent, COMSIG_LIVING_UNARMED_ATTACK, PROC_REF(owner_attacked_atom)) // This only works for players, but I am not sure this should have AI anyway
+	RegisterSignal(parent, COMSIG_ATOM_EXITED, PROC_REF(atom_exited_owner))
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS, PROC_REF(on_update_overlays))
+	ADD_TRAIT(parent, TRAIT_DISK_VERIFIER, NUKE_OP_MINION_TRAIT) // Can identify the real disk
+	ADD_TRAIT(parent, TRAIT_CAN_STRIP, NUKE_OP_MINION_TRAIT) // Can take the disk off people
+	ADD_TRAIT(parent, TRAIT_CAN_USE_NUKE, NUKE_OP_MINION_TRAIT) // Can put the disk into the bomb
+
+/datum/component/nuclear_bomb_operator/UnregisterFromParent()
+	. = ..()
+	UnregisterSignal(parent, list(COMSIG_PARENT_EXAMINE, COMSIG_LIVING_DEATH, COMSIG_PARENT_QDELETING, COMSIG_ATOM_ATTACK_HAND, COMSIG_ATOM_EXITED, COMSIG_ATOM_UPDATE_OVERLAYS))
```
Comment (Mothblocks) [inline code/datums/components/nuclear_bomb_operator.dm:43]: Should split these up into multiple lines, it's a bit hard to read right now

File: code/datums/components/nuclear_bomb_operator.dm [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+/// Drop the disk on the floor, if we have it
+/datum/component/nuclear_bomb_operator/proc/drop_disky()
+	var/obj/item/disk/nuclear/held_disk = disky?.resolve()
+	if (!held_disk)
+		return
+	var/mob/mob_parent = parent
+	held_disk.forceMove(mob_parent.drop_location())
+	mob_parent.visible_message(span_danger("[mob_parent] drops [held_disk] onto the ground!"))
+	disky = null
+	mob_parent.update_icon()
```
Comment (Mothblocks) [inline code/datums/components/nuclear_bomb_operator.dm:61]: I *think* we should be calling `update_appearance(UPDATE_ICON)` instead? It calls more signals

File: code/modules/mob/living/simple_animal/hostile/carp.dm [hunk 3]
```diff
@@ -248,78 +248,19 @@
...
-			update_icon()
-		else
-			disky.melee_attack_chain(src, attacked_target)
-		return
-
-	if(istype(attacked_target, /obj/machinery/nuclearbomb))
-		var/obj/machinery/nuclearbomb/nuke = attacked_target
-		nuke.ui_interact(src)
+	var/mutable_appearance/disk = mutable_appearance('icons/mob/simple/carp.dmi', "disk_overlay")
+	var/mutable_appearance/mouth = mutable_appearance(SSgreyscale.GetColoredIconByType(/datum/greyscale_config/carp/disk_mouth, greyscale_colors), "disk_mouth")
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/hostile/carp.dm:256]: Is there any chance that we can defer this until it's necessary without making the code hard to read? GAGS has a non-negligible impact on init times.

File: code/modules/mob/living/simple_animal/hostile/carp.dm [hunk 4]
```diff
@@ -248,78 +248,19 @@
...
+	AddComponent(/datum/component/nuclear_bomb_operator, disk, mouth)
+	RegisterSignal(src, COMSIG_HANDLESS_MOB_COLLECTED_DISK, PROC_REF(got_disk))
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/hostile/carp.dm:258]: This should be a callback parameter on the component instead of a signal

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/hostile/carp.dm (hunk 3 above)
Comment (Jacquerel) [reply code/modules/mob/living/simple_animal/hostile/carp.dm]: The only way I could think to do this is to delegate the overlays back to the mob again, as I can't think of a good way to pass GAGS info through to the component without just assembling it
Not sure if this is actually a good idea, makes it hard for someone to set up via VV too but I'm not honestly sure people doing that were going to bother giving it graphics anwyay

Comment (Mothblocks) [reply code/modules/mob/living/simple_animal/hostile/carp.dm]: Yeah I can't think of a great way myself, I'll just let this go then

File: code/datums/components/nuclear_bomb_operator.dm [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+/datum/component/nuclear_bomb_operator/proc/try_pick_up_disk(obj/item/disk/nuclear/potential_disky)
+	if(potential_disky.anchored)
+		return
+	var/mob/mob_parent = parent
+	potential_disky.forceMove(mob_parent)
+	disky = WEAKREF(potential_disky)
+	mob_parent.update_appearance(updates = UPDATE_ICON)
+	mob_parent.balloon_alert(mob_parent, "disk secured!")
+	if (on_disk_collected)
+		on_disk_collected.Invoke(potential_disky)
```
Comment (Mothblocks) [inline code/datums/components/nuclear_bomb_operator.dm:112]:
```suggestion
	on_disk_collected?.Invoke(potential_disky)
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Last few nits

File: code/datums/components/nuclear_bomb_operator.dm [hunk 6]
```diff
@@ -108,8 +108,11 @@
 	disky = WEAKREF(potential_disky)
 	mob_parent.update_appearance(updates = UPDATE_ICON)
 	mob_parent.balloon_alert(mob_parent, "disk secured!")
-	if (on_disk_collected)
-		on_disk_collected.Invoke(potential_disky)
+	INVOKE_ASYNC(src, PROC_REF(on_disk_collected), potential_disky)
```
Comment (Mothblocks) [inline code/datums/components/nuclear_bomb_operator.dm:111]: You don't need to do this, you can `on_disk_collected?.InvokeAsync(potential_disky)`

**Review by Mothblocks — APPROVED** [review]
Undraft if ready

**Pull request merged**